File: agent/src/util/categorizer.ts
// Path: agent/src/util/categorizer.ts
// Description: File kind classification (docs/code/other) based on path patterns

import { minimatch } from "minimatch";
import type { FileKind } from "../../../app/src/shared/protocol.js";

/** Extensions considered documentation */
const DOC_EXTENSIONS = new Set([
  ".md",
  ".txt",
  ".rst",
  ".adoc",
  ".asciidoc",
  ".wiki",
]);

/** Directories that indicate documentation */
const DOC_DIRS = new Set([
  "docs",
  "doc",
  "documentation",
  "wiki",
]);

/** Extensions considered code */
const CODE_EXTENSIONS = new Set([
  ".ts",
  ".tsx",
  ".js",
  ".jsx",
  ".mjs",
  ".cjs",
  ".json",
  ".rs",
  ".toml",
  ".css",
  ".scss",
  ".html",
  ".svelte",
  ".vue",
  ".py",
  ".go",
  ".yaml",
  ".yml",
]);

export interface CategorizerConfig {
  docsGlobs: string[];
  codeGlobs: string[];
}

function matchesAnyGlobs(pathName: string, globs: string[]): boolean {
  for (const glob of globs) {
    if (minimatch(pathName, glob, { dot: true, nocase: true })) {
      return true;
    }
  }
  return false;
}

function fallbackCategorize(relativePath: string): FileKind {
  const lower = relativePath.toLowerCase();
  const parts = lower.split("/");
  const fileName = parts[parts.length - 1] ?? "";

  for (const part of parts.slice(0, -1)) {
    if (DOC_DIRS.has(part)) {
      return "docs";
    }
  }

  const extMatch = /\.[^./]+$/.exec(fileName);
  const ext = extMatch?.[0] ?? "";

  if (DOC_EXTENSIONS.has(ext)) {
    return "docs";
  }

  if (CODE_EXTENSIONS.has(ext)) {
    return "code";
  }

  if (fileName === "readme" || fileName.startsWith("readme.")) {
    return "docs";
  }

  return "other";
}

/**
 * Classify a file path into docs, code, or other.
 */
export function categorizeFile(
  relativePath: string,
  config: CategorizerConfig
): FileKind {
  const normalized = relativePath.replace(/\\/g, "/");

  if (config.docsGlobs.length > 0 && matchesAnyGlobs(normalized, config.docsGlobs)) {
    return "docs";
  }

  if (config.codeGlobs.length > 0 && matchesAnyGlobs(normalized, config.codeGlobs)) {
    return "code";
  }

  return fallbackCategorize(normalized);
}

/**
 * Returns true if the file kind should be auto-staged.
 */
export function shouldAutoStage(kind: FileKind): boolean {
  return kind === "docs" || kind === "code";
}

File: crates/im_agent/src/repos/categorizer.rs
// Path: crates/im_agent/src/repos/categorizer.rs
// Description: File kind classification based on globs and fallback heuristics

use globset::{GlobBuilder, GlobSet, GlobSetBuilder};

use super::generated_code_extensions::GENERATED_CODE_EXTENSIONS;
use crate::error::AgentError;
use crate::protocol::FileKind;

const DOC_EXTENSIONS: &[&str] = &[".md", ".txt", ".rst", ".adoc", ".asciidoc", ".wiki"];
const DOC_IMAGE_EXTENSIONS: &[&str] = &[
    ".png", ".jpg", ".jpeg", ".webp", ".gif", ".bmp", ".heic", ".heif", ".tiff", ".tif",
];

const DOC_DIRS: &[&str] = &["docs", "doc", "documentation", "wiki"];

const CODE_FILE_NAMES: &[&str] = &[
    "makefile",
    "dockerfile",
    "containerfile",
    "cmakelists.txt",
    "meson.build",
    "meson_options.txt",
];

#[derive(Debug, Clone)]
pub struct Categorizer {
    docs: Option<GlobSet>,
    code: Option<GlobSet>,
}

impl Categorizer {
    pub fn new(docs_globs: &[String], code_globs: &[String]) -> Result<Self, AgentError> {
        let docs = build_glob_set(docs_globs)?;
        let code = build_glob_set(code_globs)?;

        Ok(Self { docs, code })
    }

    pub fn categorize(&self, relative_path: &str) -> FileKind {
        let normalized = relative_path.replace('\\', "/");

        if let Some(docs) = &self.docs {
            if docs.is_match(&normalized) {
                return FileKind::Docs;
            }
        }

        if let Some(code) = &self.code {
            if code.is_match(&normalized) {
                return FileKind::Code;
            }
        }

        fallback_categorize(&normalized)
    }
}

fn build_glob_set(globs: &[String]) -> Result<Option<GlobSet>, AgentError> {
    if globs.is_empty() {
        return Ok(None);
    }

    let mut builder = GlobSetBuilder::new();
    for glob in globs {
        let compiled = GlobBuilder::new(glob)
            .case_insensitive(true)
            .literal_separator(false)
            .backslash_escape(false)
            .build()
            .map_err(|err| AgentError::new("INVALID_GLOB", err.to_string()))?;
        builder.add(compiled);
    }

    builder
        .build()
        .map(Some)
        .map_err(|err| AgentError::new("INVALID_GLOB", err.to_string()))
}

fn fallback_categorize(relative_path: &str) -> FileKind {
    let lower = relative_path.to_lowercase();
    let parts: Vec<&str> = lower.split('/').collect();
    let file_name = parts.last().copied().unwrap_or_default();

    for part in parts.iter().take(parts.len().saturating_sub(1)) {
        if DOC_DIRS.iter().any(|dir| dir == part) {
            return FileKind::Docs;
        }
    }

    if let Some(ext) = file_name.rsplit_once('.') {
        let ext = format!(".{}", ext.1);
        if DOC_EXTENSIONS.contains(&ext.as_str()) {
            return FileKind::Docs;
        }
        if DOC_IMAGE_EXTENSIONS.contains(&ext.as_str()) {
            return FileKind::Docs;
        }
        if GENERATED_CODE_EXTENSIONS.contains(&ext.as_str()) {
            return FileKind::Code;
        }
    }

    if CODE_FILE_NAMES.contains(&file_name) {
        return FileKind::Code;
    }

    if file_name == "readme" || file_name.starts_with("readme.") {
        return FileKind::Docs;
    }

    FileKind::Other
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn fallback_classifies_cpp_family_as_code() {
        assert_eq!(fallback_categorize("Test.cpp"), FileKind::Code);
        assert_eq!(fallback_categorize("Source/GameMode.hpp"), FileKind::Code);
        assert_eq!(fallback_categorize("Source/LinuxEntry.cc"), FileKind::Code);
        assert_eq!(fallback_categorize("Source/Platform.mm"), FileKind::Code);
    }

    #[test]
    fn fallback_keeps_docs_priority() {
        assert_eq!(fallback_categorize("docs/notes.txt"), FileKind::Docs);
        assert_eq!(fallback_categorize("README.md"), FileKind::Docs);
    }

    #[test]
    fn fallback_classifies_images_as_docs() {
        assert_eq!(fallback_categorize("Screenshots/Capture.png"), FileKind::Docs);
        assert_eq!(fallback_categorize("captures/IMG_0001.JPG"), FileKind::Docs);
        assert_eq!(fallback_categorize("notes/snip.webp"), FileKind::Docs);
    }
}
